ONSdigital/census-contact-centre-service
# Cr 744 - create endpoint to allow Serco staff to search for a CCS case by postcode

# Motivation and Context

The CR-744 task is to create a new endpoint, in the contact centre sevice, which will allow Serco staff to search for a CCS (Census Coverage Survey) case by postcode.

Serco will not be able to find the CCS cases that RM creates using the usual UPRN endpoint; RM have implemented a new endpoint in the Case API service:

@GetMapping(value = "/ccs/postcode/{postcode}")
  public List<CaseContainerDTO> findCCSCasesByPostcode(
      @PathVariable("postcode") String postcode,
      @RequestParam(value = "caseEvents", required = false, defaultValue = "false")
          boolean caseEvents);

The new contact centre service endpoint is slightly different to the existing GET case endpoints in the CC:

- This endpoint does NOT create a cached case - it cannot as it does not have the details required
- This endpoint first checks for the existence of the requested postcode in a list of CCS postcodes and, if the postcode does not exist in that list, it responds with BAD_REQUEST.

The three possible responses from the new endpoint are:

- BAD_REQUEST : postcode not in the cached list with the message ""The requested postcode is not within the CCS sample""
- SUCCESS: RM found the CC Case by postcode
- NOT_FOUND : RM did not find the case

# What has changed

The new endpoint has been created and is named getCCSCaseByPostcode

The method signature is:

@RequestMapping(value = "/ccs/postcode/{postcode}", method = RequestMethod.GET)
  @ResponseStatus(value = HttpStatus.OK)
  public ResponseEntity<List<CaseDTO>> getCCSCaseByPostcode(
      @PathVariable(value = "postcode") @NotBlank @Pattern(regexp = Constants.POSTCODE_RE)
          final String postcode)
      throws CTPException

The endpoint validates the postcode of the request against a list of CCS postcodes, which it attempts to read in. The application yml contains the path of the postcodes file. 

The application yml also contains a list of default postcodes, which get used if the external list of postcodes is not read in successfully. However, if that happens then a warning is logged.

Seven JUnit tests have also been written for the new functionality. These are named as follows:
1. getCcsCaseByPostcode_PostcodeInList()
2. getCcsCaseByPostcode_PostcodeInListButNotInRM()
3. getCcsCaseByPostcode_PostcodeNotInList()
4. testGetCcsCaseByPostcode_caseSvcRestClientException()
5. testGetCcsCaseByPostcode_withPostcodeInRMAndInCCSPostcodes()
6. testGetCcsCaseByPostcode_withPostcodeNotInRMAndInCCSPostcodes()
7. testGetCcsCaseByPostcode_withPostcodeNotInCCSPostcodes()     

# How to test?

Run the following command to make sure that you have permission to access Firestore: gcloud auth application-default login
Run the mock case service locally (either using eclipse or mvn spring-boot:run)
Run this branch of the contact centre service.

Use Postman to send a POST request to the following URL:
http://localhost:8161/cases/data/ccs/cases/save

NB. You will need to use basic authentication with the username and password from the mock case service. The request body should be something like this:
    [{
        "caseRef": "123123002",
        "estabType": "ET",
        "uprn": "1347459999",
        "createdDateTime": "2019-04-14T13:45:26.564+01:00",
        "addressLine1": "Happy House",
        "addressLine2": "89 Harbour Street",
        "addressLine3": "Somewhere",
        "townName": "Portsmouth",
        "postcode": "GW12 AAA",
        "organisationName": "ON",
        "addressLevel": "E",
        "abpCode": "AACC",
        "latitude": "41.40338",
        "longitude": "2.17403",
        "oa": "EE22",
        "lsoa": "x1",
        "msoa": "x2",
        "lad": "H1",
        "caseEvents": [
            {
                "id": "101",
                "eventType": "CASE_UPDATED",
                "description": "Initial creation of case",
                "createdDateTime": "2019-04-14T13:45:26.564+01:00"
            },
            {
                "id": "102",
                "eventType": null,
                "description": "Create Household Visit",
                "createdDateTime": "2019-04-14T13:45:26.564+01:00"
            }
        ],
        "id": "3305e937-6fb1-4ce1-9d4c-077f147789aa",
        "caseType": "HH",
        "addressType": "SPG",
        "region": "E",
        "collectionExerciseId": "49871667-117d-4a63-9101-f6a0660f73f6",
        "surveyType": "CENSUS",
        "handDelivery": false
    }]

You can now use the cc svc endpoint to check that you get back the three expected responses, as follows:

1. This URL should give a successful outcome, which is to say that it should return the above case that you've added to the mock case service:

http://localhost:8171/cases/ccs/postcode/GW12 AAA

2. This URL should give a not found message:

http://localhost:8171/cases/ccs/postcode/GW12 AAB

3. This URL should give a bad request message with the text "The requested postcode is not within the CCS sample":

http://localhost:8171/cases/ccs/postcode/GW12 AAC
